src/openvic-simulation/GameManager.cpp

```diff
@@ -0,0 +1,132 @@
+#include "GameManager.hpp"
+
+#include "openvic-simulation/utility/Logger.hpp"
+
+using namespace OpenVic;
+
+GameManager::GameManager(state_updated_func_t state_updated_callback)
+	: clock { [this]() { tick(); }, [this]() { update_state(); } },
+	  state_updated { state_updated_callback } {}
+
+void GameManager::set_needs_update() {
+	needs_update = true;
+}
+
+void GameManager::update_state() {
+	if (needs_update) {
+		Logger::info("Update: ", today);
+		map.update_state(today);
+		if (state_updated) state_updated();
+		needs_update = false;
+	}
+}
+
+/* REQUIREMENTS:
+ * SS-98, SS-101
+ */
+void GameManager::tick() {
+	today++;
+	Logger::info("Tick: ", today);
+	map.tick(today);
+	set_needs_update();
+}
+
+bool GameManager::setup() {
+	session_start = time(nullptr);
+	clock.reset();
+	today = { 1836 };
+	good_manager.reset_to_defaults();
+	bool ret = map.setup(good_manager, building_manager, pop_manager);
+	set_needs_update();
+	return ret;
+}
+
+Date const& GameManager::get_today() const {
+	return today;
+}
+
+bool GameManager::expand_building(Province::index_t province_index, const std::string_view building_type_identifier) {
+	set_needs_update();
+	Province* province = map.get_province_by_index(province_index);
+	if (province == nullptr) {
+		Logger::error("Invalid province index ", province_index, " while trying to expand building ", building_type_identifier);
+		return false;
+	}
+	return province->expand_building(building_type_identifier);
+}
+
+bool GameManager::load_hardcoded_defines() {
+	bool ret = true;
+
+	static constexpr colour_t LOW_ALPHA_VALUE = float_to_alpha_value(0.4f);
+	static constexpr colour_t HIGH_ALPHA_VALUE = float_to_alpha_value(0.7f);
+	using mapmode_t = std::pair<std::string, Mapmode::colour_func_t>;
+	const std::vector<mapmode_t> mapmodes {
```

Comment:
This isn't really an issue, but this is probably another good place to suggest the incorporating an ordered map of contiguous memory, specifically ordered on insertion order.

Reply:
These are "hardcoded defines", which ideally could be replaced with `txt` defines once we have a powerful enough scripting format. The vector is iterated over with the `std::string` and `Mapmode::colour_func_t` passed to `map.add_mapmode`, which validates the arguments, constructs a `Mapmode` (move-only, like its base-class `HasIdentifier`) and enters it into an `IdentifierRegistry<Mapmode>`. There, `Mapmode`s are stored in a `std::vector` to allow lookup by index, with a `std::map<std::string, size_t, std::less<void>>` used for lookup by identifier (the `std::less<void>` allows us to search by `std::string_view`).  Feel free to share any ideas for improvements,